### PR TITLE
Fix children no reordered after zIndex and sibling index change in Node

### PR DESCRIPTION
**Pull request by pandamicro**
> Makes sure these boxes are checked before submitting your PR - thank you!
> - [x] Check that your javascript is following our [style guide](https://github.com/cocos-creator/fireball/blob/dev/.github/CONTRIBUTING.md) and end files with a newline
> - [x] Document new code with comments in source code based on [API Docs](https://github.com/cocos-creator/fireball#api-docs)
> - [x] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
> - [x] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.

Re: cocos-creator/fireball#2477

Changes proposed in this pull request:
- 修复 children 数组在 zIndex 和 sibling index 修改后没有改变的问题

@cocos-creator/engine-admins

### CODE REVIEW

File: cocos2d/core/utils/base-node.js [hunk 1]
```diff
@@ -247,6 +248,12 @@ var BaseNode = cc.Class(/** @lends cc.Node# */{
             set: function (value) {
                 this._localZOrder = value;
                 this._sgNode.zIndex = value;
+
+                if (!CC_JSB && this._parent) {
+                    this._parent._reorderChildDirty = true;
```
Comment (jareguo) [inline cocos2d/core/utils/base-node.js:253]: _reorderChildDirty 目前没定义，建议设置到 _dirtyFlags 里面

File: cocos2d/core/utils/base-node.js [hunk 2]
```diff
@@ -1782,9 +1789,11 @@ var BaseNode = cc.Class(/** @lends cc.Node# */{
                 sibling._sgNode.arrivalOrder = i;
             }
             if ( !CC_JSB ) {
-                cc.renderer.childrenOrderDirty = this._parent._sgNode._reorderChildDirty = true;
+                cc.renderer.childrenOrderDirty = true;
+                this._parent._sgNode._reorderChildDirty = true;
+                this._parent._reorderChildDirty = true;
+                cc.director.once(cc.Director.EVENT_AFTER_UPDATE, this._parent.sortAllChildren, this._parent);
             }
-            this._parent.emit(CHILD_REORDER);
```
Comment (jareguo) [inline cocos2d/core/utils/base-node.js]: 这个事件 layout 有用到，不应该去掉

File: cocos2d/core/utils/base-node.js [hunk 3]
```diff
@@ -1782,9 +1789,11 @@ var BaseNode = cc.Class(/** @lends cc.Node# */{
                 sibling._sgNode.arrivalOrder = i;
             }
             if ( !CC_JSB ) {
-                cc.renderer.childrenOrderDirty = this._parent._sgNode._reorderChildDirty = true;
+                cc.renderer.childrenOrderDirty = true;
+                this._parent._sgNode._reorderChildDirty = true;
+                this._parent._reorderChildDirty = true;
+                cc.director.once(cc.Director.EVENT_AFTER_UPDATE, this._parent.sortAllChildren, this._parent);
```
Comment (jareguo) [inline cocos2d/core/utils/base-node.js]: this.sortAllChildren 就行了

File: cocos2d/core/utils/base-node.js (hunk 1 above)
Comment (pandamicro) [inline cocos2d/core/utils/base-node.js:253]: 这个是 Node 级别的 dirty，并没有 sgNode 那样的 dirty 机制，加进去也不太合适

File: cocos2d/core/utils/base-node.js (hunk 3 above)
Comment (pandamicro) [inline cocos2d/core/utils/base-node.js]: 如果在同一帧内在同一个父节点的内部对多个子节点调用 setSiblingIndex，会导致重复进行排序

**Pull request merged**